File: src/commands/imagine.rs
use crate::utils::gemini_imagen_client::GeminiImagenPrompt;
use poise::serenity_prelude::{Color, CreateAttachment, CreateEmbed};
use poise::{command, CreateReply};
use tracing::info;

type Error = Box<dyn std::error::Error + Send + Sync>;
type Context<'a> = poise::Context<'a, crate::Data, crate::Error>;

#[command(
  slash_command,
  description_localized("en-US", "Image generation with slowpoke")
)]
pub async fn imagine(
  ctx: Context<'_>,
  #[description = "Prompt for image generation"] prompt: String,
) -> Result<(), Error> {
  // As AI models take some time to respond, defer must be called first in order to keep
  // the interaction alive.
  ctx.defer().await?;

  let gemini_imagen_client = &ctx.data().gemini_imagen_client;
  let imagen_prompt = GeminiImagenPrompt {
    prompt: prompt.clone(),
  };

  let image_data = match gemini_imagen_client.prompt(imagen_prompt).await {
    Ok(generated_image) => generated_image.image_data,
    Err(e) => {
      info!("Error imagining image: {:?}", e);
      let embed = CreateEmbed::default()
        .title("Error imagining image")
        .description(
          "Failed to imagine image. Try altering the prompt or trying again later.",
        )
        .color(Color::from_rgb(137, 220, 235));
      ctx
        .send(CreateReply::default().embed(embed).reply(true))
        .await?;
      return Ok(());
    }
  };

  let attachment = CreateAttachment::bytes(image_data, "image.png");

  let embed = CreateEmbed::default()
    .title("Imagine")
    .description(prompt)
    .color(Color::from_rgb(137, 220, 235))
    .attachment(&attachment.filename);

  ctx
    .send(CreateReply::default().embed(embed).attachment(attachment))
    .await?;

  Ok(())
}

File: src/commands/imagine.ts
import { SlashCommandBuilder, ChatInputCommandInteraction, EmbedBuilder, AttachmentBuilder } from 'discord.js'
import { GeminiClient } from '../utils/gemini-client'
import { logger } from '../utils/logger'
import { SlashCommand } from '../models/commands'
import { ColorProvider } from '../utils/color-provider'

export const createImagineCommand = (geminiClient: GeminiClient, colorProvider: ColorProvider): SlashCommand => ({
  command: new SlashCommandBuilder()
    .setName('imagine')
    .setDescription('Image generation with slowpoke')
    .addStringOption(option =>
      option
        .setName('prompt')
        .setDescription('Prompt for image generation')
        .setRequired(true)
    ),

  async execute (interaction: ChatInputCommandInteraction) {
    // Defer the reply as AI models take time to respond
    await interaction.deferReply()

    const promptText = interaction.options.get('prompt')?.value as string

    try {
      const imageData = await geminiClient.generateImage({
        prompt: promptText
      })

      const attachment = new AttachmentBuilder(imageData, { name: 'image.png' })

      const embed = new EmbedBuilder()
        .setTitle('Imagine')
        .setDescription(promptText)
        .setColor(colorProvider.getPrimaryColor())
        .setImage('attachment://image.png')

      await interaction.editReply({
        embeds: [embed],
        files: [attachment]
      })
    } catch (error) {
      logger.error({ error }, 'Error imagining image')

      const errorEmbed = new EmbedBuilder()
        .setTitle('Error imagining image')
        .setDescription('Failed to imagine image. Try altering the prompt or trying again later.')
        .setColor(colorProvider.getErrorColor())

      await interaction.editReply({ embeds: [errorEmbed] })
    }
  }
})
